### PR TITLE
feat: support sandbox mode

### PR DESCRIPTION
**Pull request by lowlighter**
Towards #33

This adds a `sandbox` parameters to `browser.newPage()`:

- If `sandbox: true`, each network request is intercepted and checked against deno permissions
- If `sandbox: Network_ResourceType[]`, only network requests matching the resource type are intercepted and checked against deno permissions (for example, you can choose to just sandbox `document` or `fetch`  request)

Unfortunately because of https://github.com/denoland/deno/issues/14668, I had to add a `sandboxPrompt` arg to toggle interactivity, since the current `Deno.permissions.request` will always ask for permissions even when passing `--no-prompt` or with `DENO_NO_PROMPT=1`

From the unit test I added, it seems pretty robuts, as you even if you try to use `page.evaluate` to redirect or fetch with JS, the network request will be intercepted (provided you pass the correct resource type in sandbox)

Small demo:
ezgif-1-3ad9a18719

Side notes:
- For the tests, I use `window.performance.getEntriesByType('navigation')[0].responseStatus` which returns the status code of the page. When using `#celestial.Fetch.failRequest` the response status will be `0` **however** the page is not empty, from I've seen it display a chrome error page telling that the request failed.
- Additional thoughts: should `file://` protocol use `--allow-read` permissions instead and supported by this feature ?

### CODE REVIEW

File: src/page.ts
Comment (lino-levan) [inline src/page.ts:97]: Couldn't this just be
```suggestion
    options: SandboxOptions,
```

File: src/page.ts
Comment (lino-levan) [inline src/page.ts:129]: This creates a race condition (which is unlikely to matter in IRL code, but I've encountered before unfortunately). Could you move the fetch enabling to the browser's `newPage` method? I believe there should be some other enablings there.

File: src/page.ts
Comment (lino-levan) [inline src/page.ts:47]: Is there a reason that a user would only want to sandbox certain types of request? What's the usecase here? Wouldn't this be a security foot-gun?

File: src/page.ts
Comment (lino-levan) [inline src/page.ts:50]: Could you add JSDocs explaining what these options do?

File: src/page.ts
Comment (lino-levan) [inline src/page.ts:7]: Maybe these should be imported as types instead of really being imported? I think it was a mistake of mine to not do that in the first place.
```suggestion
import type {
  Celestial,
  Network_Cookie,
  Network_ResourceType,
} from "../bindings/celestial.ts";
```

File: src/page.ts
Comment (lino-levan) [inline src/page.ts:137]: Are these always absolute URLs? If not, we should set the base.

**Review by lino-levan — CHANGES_REQUESTED** [review]
This is such a cool feature! Thanks for picking it up. I love the integration with Deno's permission system.

**Comment by lino-levan** [comment]
> Unfortunately because of https://github.com/denoland/deno/issues/14668, I had to add a sandboxPrompt arg to toggle interactivity, since the current Deno.permissions.request will always ask for permissions even when passing --no-prompt or with DENO_NO_PROMPT=1

I'll just fix upstream, no need to add yet another option. Not a super hard patch.

File: src/page.ts
Comment (lowlighter) [reply src/page.ts:129]: Ok I'll move it this to `browser.ts` instead 👍

File: src/page.ts
Comment (lowlighter) [reply src/page.ts:137]: I need to check as the [`Network.Request`](https://chromedevtools.github.io/devtools-protocol/tot/Network/#type-Request) doc doesn't precise, but think they are

File: src/page.ts
Comment (lowlighter) [reply src/page.ts:50]: I'll add the docs once it's decided what API should be adopted 👍

File: src/page.ts
Comment (lowlighter) [reply src/page.ts:47]: I think it depends on whether you want granularity at "website level" or "asset level"

My use-case would be the former, where I'd just like to control what website can be accessed, but don't really care where they need to load assets from (like if they use external domains from CDN or hotlinks).
I guess it'd limit my checks to `Document` requests, as for me static assets like `Stylesheet`, `Image`, `Media`, `Font` or  `Script` would be fine as I'd consider I already "trusted" the document itself.

But to be honest the [`Network.ResourceType`](https://chromedevtools.github.io/devtools-protocol/tot/Network/#type-ResourceType) has a lot of types (Document, Stylesheet, Image, Media, Font, Script, TextTrack, XHR, Fetch, Prefetch, EventSource, WebSocket, Manifest, SignedExchange, Ping, CSPViolationReport, Preflight, Other) and I'm not sure what all of them corresponds too (like `EventSource`, `SignedExchange`, `Preflight` or `Other`).

I feel like the current API I offer in this PR exposes a bit too much stuff, but I really think we should have something to address the "website level" or "asset level" control. 
I can see where both could be useful, like if you're testing your own website maybe you want to be sure that you load nothing suspicious, but if you're just webscraping stuff, maybe you just want to restrict which website can be accessed but you may not know in advance where the website will load its content

**Comment by lowlighter** [comment]
Thanks for the quick review and the feedback !
I may follow this pr a bit sparsely since I'm doing other stuff in parallel but I'll try to be responsive on it 😄

**Comment by lino-levan** [comment]
Thank you for your work! Take your time.

File: src/page.ts
Comment (lowlighter) [reply src/page.ts]: Finally decided that it's better to just leave a boolean for `sandbox` to make it less complex.
If it's too limiting I guess it can be improved later on

File: src/page.ts
Comment (lowlighter) [reply src/page.ts]: After playing a bit with this, seems it's always absolute (like `fetch("/test")` will be resolved related to the current domain page)

File: src/page.ts
Comment (lino-levan) [inline src/page.ts:155]: Does this make sense? When does this condition trigger?

File: src/page.ts
Comment (lino-levan) [reply src/page.ts]: Agree on the decision. I think we could formulate a less leaky abstraction at some point in the future if a need is discovered.

**Comment by lino-levan** [comment]
Any ideas why tests are failing?

File: src/page.ts
Comment (lowlighter) [reply src/page.ts:155]: I don't think it's possible to reach the `return true` statement, but in the case that it could exist requests that are not using `file:` protocol and don't have a host then they'd be systematically rejected if nothing was returned.
AFAIK it doesn't exist, but maybe there are some edge cases.

As a side note, I did try using `data:` scheme and it totally bypass the sandbox as it does not trigger the `Fetch` namespace at all when passed

**Comment by lowlighter** [comment]
> Any ideas why tests are failing?

For some reason this test does not work on windows (is ok on linux/mac os):
```ts
Deno.test.ignore("Sandbox fulfill granted read permissions", {
  permissions: {
    ...permissions,
    read: [...permissions.read, fromFileUrl(import.meta.url)],
    net: ["127.0.0.1"],
  },
}, async () => {
  const browser = await launch();
  const page = await browser.newPage(import.meta.url, { sandbox: true });
  assertStrictEquals(await page.evaluate(status), 200);
  await browser.close();
});
```
I'll try to debug this when I have some time. I wonder if it's because the volume name on windows make the url weird or something similar.

Btw, I think the error code 21 is actually the "lock file" of windows (since windows can create lock directly on the files). I think that's why if one of the tests fail on windows, the whole suite will start to behave weirdly and fail too, because `browser.close()` is called after the assertions so if one of them throw the process is actually never closed properly and it's not possible to relaunch an instance

File: src/page.ts
Comment (lino-levan) [reply src/page.ts:157]: Good to know, thanks for investigating. We can change later if we realize that this is not correct.

**Review by lino-levan — APPROVED** [review]
Any changes you want to make before this gets merged? This LGTM. Great work!

**Comment by lino-levan** [comment]
> Btw, I think the error code 21 is actually the "lock file" of windows

I think that makes a lot of sense actually. Could definitely be the case.

**Comment by lowlighter** [comment]
> Any changes you want to make before this gets merged? This LGTM. Great work!

Apart from changing `Deno.permissions.query` to `Deno.permissions.request` once it's fixed upstream not really.

For now it means that enabling sandbox will not prompt for additional permissions so users shouldn't be surprised if their requests fail if they didn't grant enough permissions

**Comment by lino-levan** [comment]
I think this is a reasonable limitation for now. We can change it to prompt when I patch upstream. Thanks for your work!

**Pull request merged**
